lib/Sema/TypeCheckDecl.cpp

```diff
@@ -2757,7 +2757,17 @@ void swift::markAsObjC(TypeChecker &TC, ValueDecl *D,
       TC.diagnose(D, diag::objc_inference_swift3, *deprecatedReason)
         .highlight(dynamicLoc)
         .fixItInsert(D->getAttributeInsertionLoc(false), "@objc ");
-    };
+    }
+
+    // Mark the attribute as having used Swift 3 inference, or create an
+    // implicit @objc for that purpose.
+    auto attr = D->getAttrs().getAttribute<ObjCAttr>();
+    if (!attr) {
```

Comment:
Ah, you're right! I'd forgotten that we fully rely on the presence of `ObjCAttr`.